cf/query.py

```diff
@@ -790,7 +791,11 @@ def set_condition_units(self, units):
         value_units = getattr(value, "Units", None)
         if value_units is None:
             # Value has no units
-            value = Data(value, units=units)
+            if isinstance(value, Iterable):  # may be a sequence of Data
```

Comment:
Hi Sadie, not sure this will work in all cases:

* `Data` objects are `Iterable`
* When `value` is a sequence, we want it to remain a sequence, not convert it into a `Data` object, and `Data.concatenate` doesn't work for non-Data objects 

This code might make sense, but it's not very pretty!
```python
        value_units = getattr(value, "Units", None)
        if value_units is None:
            # Value has no units
            if isinstance(value, Iterable) and not isinstance(value, str):
                new = []
                for v in value:
                    value_units = getattr(v, "Units", None)
                    if value_units is None:
                        # Value has no units
                        v = Data(v, units=units)
                    else:
                        # Value already has units
                        try:
                            v.Units = units
                        except ValueError:
                            raise ValueError(
                                f"Units {units!r} are not equivalent to "
                                f"query condition units {value_units!r}"
                            )
                        
                    new.append(v)

                value = new
            else:
                value = Data(value, units=units)        
        else:
            # Value already has units
            try:
                value.Units = units
            except ValueError:
                raise ValueError(
                    f"Units {units!r} are not equivalent to "
                    f"query condition units {value_units!r}"
                )
```

Reply:
OK - I think I worked out what's going on. The key is that the contents of a "set", "wi" or "wo" iterable are not necessarily `Data` objects (or rather objects with `Units`. How about this? There's scope for a bit of code re-use here, but I can't quite see how, yet.

```python

        value_units = getattr(value, "Units", None)
        if value_units is None:
            # Value has no units
            if self.operator in ("wi", "wo", "set"):
                # value is a sequence of things that may or may not
                # already have units
                new = []
                for v in value:
                    v_units = getattr(v, "Units", None)
                    if v_units is None:
                        v = Data(v, units=units)
                    else:
                        try:
                            v = v.copy()
                            v.Units = units
                        except ValueError:
                            raise ValueError(
                                f"Units {units!r} are not equivalent to "
                                f"query condition units {v_units!r}"
                            )
                        
                    new.append(v)

                value = new
            else:
                value = Data(value, units=units)
        else:
            # Value already has units
            try:
                value = value.copy()
                value.Units = units
            except ValueError:
                raise ValueError(
                    f"Units {units!r} are not equivalent to "
                    f"query condition units {value_units!r}"
                )

        self._value = value
```
We could do with some more units test for this, too ...

Reply:
I've just added your code and then consolidated it with an inner helper function. I'll add some new testing first thing tomorrow.